### ISSUE TITLE
YoloV3 one class, Trained by Darknet produces massively different mAP after convert.

### ISSUE THREAD

**Issue by mvanlierBCG** [david8862/keras-YOLOv3-model-set#41]
I have trained a default YoloV3 (COCO pre-trained) with the original Darknet Framework, on a custom (one class) dataset. The only changes of the default YoloV3 are the no. of classes, as provided below.

With a predefined test-set, with Darknet an mAP of 47.9 is achieved (IoU=0.5, thres=0.25).
The Darknet model weights and cfg is converted with the `tools\convert.py`. But when the model is evaluated on the same test-set, with the `eval.py` I am only able to achieve mAP of 12.3, and mPrec of 52.1 (IoU=0.5, thres=0.25).

Since this is a massive difference of tens of mAPs (and not just a few), my guess is that there should be a bug somewhere in the process: weight conversion, running Yolo, or evaluation.

I have done some troubleshooting: 

- Default Yolov3 COCO is tested correctly on the COCO dataset with an mAP close to the reported one.

- the annotation files are identical, (both Darknet/Keras) 

- For the weight conversion, I am not sure (outputs of the models are: 1x1x1024x18, 1x1x512x18, 1x1x256x18, which looks correct).

- Converting weights only, and then loading a default yoloV3 with Darknet 53 backbone, and loading the weights in the created skeleton, but this didn't make a difference.

- I have used mmDNN [(](https://github.com/microsoft/MMdnn) for converting, but this resulted in unknown layer(s).

- One thing I noticed but I am not sure is the `letterbox_resize()` function. I have mainly square images (500x500) +/- 15% of the images, and extreme rectangles (2452x1030), while the input of Yolo is the default (608x608). When I look at the output of the `eval.py`, in `results/detection` folder (imges with bboxes) with the Darknet evaluation (images with bboxes), it visually looks if the bboxes in the square imgs (500x500) are exactly the same, while the rectangle images usually misses a lot of bboxes, and ar usually unaligned. So the only reason that I can imagine is that the letterbox_resize causes a lot of distortion in the images that should be massively resized, which causes this problem.

I am totally lost how to solve this problem which is very important to me. It is very strange, especially since converting a default YoloV3 COCO with 80 classes works, but converting it to 1 class and a custom dataset doesn't.

**Comment by david8862** [comment]
Hi @mvanlierBCG , `letterbox_resize` is to padding the origin images to target square model-input size (e.g. 416x416) and kept the aspect ratio, like following:

![1](https://user-images.githubusercontent.com/25219223/81691374-4c538e80-948f-11ea-90ac-e36a316e85f0.jpg)

So one possible debugging is manually resize your test images to target input size (416x416) and feed to the model (that will actually bypass the letterbox_resize), then check the result in `results/detection` to see if the visual bboxes are correct

**Comment by mvanlierBCG** [comment]
thanks for your suggestion. It seems that doesn't solve the problem. It looks like most of the detection is misaligned. When I reduce the IoU, then my mAP is massively increasing (e.g. 0.3, then mAP=33) and becoming more close to the Darknet eval (47mAP). I think the reason that the objects are correctly detected in the small square images (500x500) is that the objects usually take between 20-80% of the total image, and will always have a higher IoU, even when estimated bbox is unlined. While the bigger images in my test-set (2452x1030) the objects take most of the time between 0.001 up to 0.1% of the image space, and therefore when the IoU will always be lower or close to zero just with a small estimated bbox unalignment.

For this reason, I think my suggestion on the `letterbox_resize()` was not correct.

I am now totally stuck on how I could solve this problem, and where to search for a solution. Do you have any suggestions? Could it be, e.g. that the model/weight converter is not correctly converting the weights? Eventually, I could share my model and test-set + Darknet golden master.

**Comment by david8862** [comment]
Sounds really weired. I can take a look if having trained model and testset

**Comment by mvanlierBCG** [comment]
I am totally lost indeed. What is the best/usual way to share this with you? An example could be e.g. with Google Drive.

**Comment by david8862** [comment]
Google Drive is ok for me.
One more thing is what's your anchor used in darknet and keras? If the anchor boxes for low level feature map (e.g 76x76) are different, it will impact inference result for small objects.

**Comment by mvanlierBCG** [comment]
The anchor used in Darknet and Keras are unchanged, and just the default yolov3 ones.
